lib/ui/painting/image.h

```diff
@@ -46,7 +48,7 @@ class CanvasImage final : public RefCountedDartWrappable<CanvasImage> {
  private:
   CanvasImage();
 
-  flutter::SkiaGPUObject<SkImage> image_;
+  sk_sp<DlImage> image_;
```

Comment:
I'm concerned that we moved the SkiaGPUObject wrapper down to an unnecessary level.

That wrapper exists because lib/ui/painting/image.h lives in a Dart wrapper and its lifetime is determined by Dart and its GC in a Dart isolate/thread. That isolate/thread doesn't have the ability to free things like images and so the SkiaGPUObject wrapper forwards the unref to a thread that can.

From this object, that forwarding is needed.

But, when a reference from this field is handed off to the engine, it is managed by a thread that can unref and free images, so it no longer needs the SkiaGPUObject protection. You've just moved that indirection down into the engine object which means that every time we get rid of one of these DlImage wrappers, we'll have to forward that unref to a thread.

Shouldn't this be held in a SkiaGPUObject here and then the DlObject you create here doesn't need to use a SkiaGPUObject?

Reply:
Using a SkiaGPUObject here would mean that Impeller textures would require an unnecessary thread hop on each collection. Impeller has no restriction on where textures may be collected and it (thread hops) seemed like something that ought to be left to the backend and not performed by default.